package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,468 @@
+                    # store set of atom ids which satisfy selection
+                    set_list[i].append(set(temp_list))
+                    # link atom id to array index
+                    temp_dict = dict()
+                    for j in range(len(temp_list)):
+                        temp_dict[temp_list[j]] = j
+                    dict_list[i].append(temp_dict)
+                else:
+                    pos[i].append(np.array([]))
+                    set_list[i].append(set())
+                    dict_list[i].append(dict())
+
+    def _process_pos_data(self, pos, set_list, dict_list):
+        """Runs MSD calculation and returns data
+
+        Parameters:
+        -----------
+        pos : list
+            contains position information, organized:
+            selection X frame X atom X (x, y, z)
+        set_list : list
+            contains set of atoms which satisfy each selection at each
+            frame, organized:
+            selection X frame X set of atom ids
+        dict_list : list
+            contains dictionaries which map atom ids to corresponding
+            pos index, organized:
+            selection X frame X dict
+
+        Returns:
+        --------
+        msd : np.array(dtype=float32)
+            Contains msd data, organized:
+            selection X tau
+        n_samples : np.array(dtype=int)
+            Contains the number of samples corresponding to each msd
+            entry, organized:
+            selection X tau
+        """
+        n_sel = len(pos)
+
+        n_samples = np.zeros((n_sel, self.len_msd), dtype='int')
+        msd = np.zeros((n_sel, self.len_msd), dtype='float32')
+        
+        calc_cutoff = self.tf
+        if self.max_data:
+            calc_cutoff = len(self.universe.trajectory) - 1
+
+        # for each restart point
+        for j in range(self.t0, self.tf+1, self.dt_restart):
+            for i in range(n_sel):  # for each selection
+                atoms_in_sel = set_list[i][0]  # storing initial set at restart
+                for ts in range(j, calc_cutoff+1):  # for each frame after restart
+                    # avoid computing irrelevantly long taus
+                    if ts-j == self.len_msd:
+                        break
+
+                    # updating restart set to exclude any atoms which have
+                    # left the selection since the restart point
+                    atoms_in_sel = atoms_in_sel.intersection(set_list[i][ts-j])
+                    
+                    # find mutual atoms at times 0 and ts-j
+                    shared0 = [dict_list[i][0][k] for k in atoms_in_sel]
+                    shared = [dict_list[i][ts-j][k] for k in atoms_in_sel]
+                    
+                    # move to next restart if there's nothing to evaluate
+                    if len(shared) == 0:
+                        break  # skip to next restart
+                    
+                    msd[i][ts-j] = (msd[i][ts-j]*n_samples[i][ts-j] +
```

Comment:
This would be a nice optimisation, but I'm happy to let this PR use this method for now